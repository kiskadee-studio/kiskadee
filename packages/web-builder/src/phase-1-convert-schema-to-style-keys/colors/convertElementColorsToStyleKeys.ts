import type {
  SemanticColor,
  InteractionState,
  ColorProperty,
  StyleKeyByElement,
  PaletteName,
  ElementColors
} from '@kiskadee/schema';
import { buildStyleKey, deepUpdate } from '../../utils';

/**
 * Converts an element's color palettes schema into nested style keys.
 *
 * Processes each palette and color property, handling both raw InteractionStateColorMap
 * and semantic color groups, and generates style keys for each interaction state,
 * including reference indicators, via {@link buildStyleKey}.
 *
 * @param palettes - The ElementColors object defining element palettes.
 * @returns A nested map of style keys organized by palette name and property.
 */
export function convertElementColorsToStyleKeys(
  palettes: ElementColors
): StyleKeyByElement['palettes'] {
  const styleKeys: StyleKeyByElement['palettes'] = {};

  for (const p in palettes) {
    const paletteName = p as PaletteName;
    const colorSchema = palettes[paletteName];

    for (const c in colorSchema) {
      const colorProperty = c as ColorProperty;
      const colorEntry = colorSchema[colorProperty];
      if (colorEntry === undefined) continue;

      const isInteractionState = 'rest' in colorEntry;
      const semanticColorMap = isInteractionState ? { neutral: colorEntry } : colorEntry;

      for (const s in semanticColorMap) {
        const semanticColor = s as SemanticColor;
        const interactionStateMap = semanticColorMap[semanticColor];
        for (const i in interactionStateMap) {
          const interactionState = i as InteractionState;
          const colorValue = interactionStateMap[interactionState];
          const isRef = typeof colorValue === 'object' && 'ref' in colorValue;
          const color = JSON.stringify(isRef ? colorValue.ref : colorValue);

          const styleKey = buildStyleKey({
            propertyName: colorProperty,
            interactionState: interactionState,
            isRef,
            value: color
          });

          deepUpdate(
            styleKeys,
            [paletteName, semanticColor, interactionState],
            (arr: string[] = []) => [...arr, styleKey]
          );
        }
      }
    }
  }

  return styleKeys;
}
