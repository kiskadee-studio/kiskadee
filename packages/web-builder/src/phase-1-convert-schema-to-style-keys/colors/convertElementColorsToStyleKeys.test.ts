import type { ElementPalettes } from '@kiskadee/schema';
import { describe, expect, it } from 'vitest';
import { convertElementColorsToStyleKeys } from './convertElementColorsToStyleKeys';

/*------------------------------------------------------------------------------------------------*/
/* There is no error handling here; errors were handled during the style key to CSS conversion in
/*   phase 4
/*------------------------------------------------------------------------------------------------*/

describe('convertElementColorsToStyleKeys', () => {
  it('generates style keys for palette property without reference', (): void => {
    const elementPalettes: ElementPalettes = {
      ios: {
        light: {
          boxColor: {
            primary: {
              solid: { rest: [45, 100, 50, 1] }
            }
          }
        }
      }
    };
    const result = convertElementColorsToStyleKeys(elementPalettes);
    expect(result.styleKeys).toEqual({
      ios: {
        light: {
          primary: {
            rest: ['boxColor__[45,100,50,1]']
          }
        }
      }
    });
  });

  it('generates style keys for palette property with a reference value', (): void => {
    const elementPalettes: ElementPalettes = {
      ios: {
        light: {
          borderColor: {
            primary: {
              solid: {
                rest: [255, 255, 255, 1],
                hover: { ref: [255, 255, 255, 0.1] }
              }
            }
          }
        }
      }
    };
    const result = convertElementColorsToStyleKeys(elementPalettes);
    expect(result.styleKeys).toEqual({
      ios: {
        light: {
          primary: {
            rest: ['borderColor__[255,255,255,1]'],
            hover: ['borderColor==hover__[255,255,255,0.1]']
          }
        }
      }
    });
  });

  it('generates style keys for multiple palette entries', (): void => {
    const elementPalettes: ElementPalettes = {
      ios: {
        light: {
          textColor: {
            primary: {
              solid: {
                rest: [120, 50, 50, 1],
                hover: { ref: [240, 50, 50, 0.5] }
              }
            },
            secondary: {
              solid: {
                rest: [240, 50, 50, 0.5]
              }
            }
          },
          borderColor: {
            primary: {
              solid: {
                rest: [120, 50, 50, 1]
              }
            },
            redLike: {
              solid: {
                rest: [0, 0, 0, 0.02],
                focus: { ref: [10, 20, 30, 0.1] }
              }
            }
          }
        }
      }
    };
    const result = convertElementColorsToStyleKeys(elementPalettes);
    expect(result.styleKeys).toEqual({
      ios: {
        light: {
          primary: {
            rest: ['textColor__[120,50,50,1]', 'borderColor__[120,50,50,1]'],
            hover: ['textColor==hover__[240,50,50,0.5]']
          },
          secondary: {
            rest: ['textColor__[240,50,50,0.5]']
          },
          redLike: {
            rest: ['borderColor__[0,0,0,0.02]'],
            focus: ['borderColor==focus__[10,20,30,0.1]']
          }
        }
      }
    });
  });

  it('throws when using legacy direct interaction-state map at property root (no soft/solid)', (): void => {
    const elementPalettes: ElementPalettes = {
      ios: {
        light: {
          // Legacy direct interaction‐state map at property level (invalid now)
          boxColor: {
            rest: [0, 128, 255, 1],
            hover: { ref: [0, 128, 255, 0.5] }
            // biome-ignore lint/suspicious/noExplicitAny: required for supporting legacy format
          } as any
        }
      }
    };
    expect(() => convertElementColorsToStyleKeys(elementPalettes)).toThrowError(
      /no longer supported|must define soft\/solid/i
    );
  });

  it('handles selected submap: emits selected/rest and selected:hover keys', (): void => {
    const elementPalettes: ElementPalettes = {
      ios: {
        light: {
          boxColor: {
            primary: {
              solid: {
                rest: [10, 20, 30, 0.9],
                hover: [15, 25, 35, 0.9],
                selected: {
                  rest: [200, 50, 50, 1],
                  hover: { ref: [210, 55, 55, 0.8] }
                },
                disabled: [0, 0, 50, 0.5]
              }
            }
          }
        }
      }
    };

    const result = convertElementColorsToStyleKeys(elementPalettes);

    expect(result.styleKeys).toEqual({
      ios: {
        light: {
          primary: {
            rest: ['boxColor__[10,20,30,0.9]'],
            hover: ['boxColor--hover__[15,25,35,0.9]'],
            'selected:rest': ['boxColor--selected:rest__[200,50,50,1]'],
            'selected:hover': ['boxColor==selected:hover__[210,55,55,0.8]'],
            disabled: ['boxColor--disabled__[0,0,50,0.5]']
          }
        }
      }
    });
  });
});
