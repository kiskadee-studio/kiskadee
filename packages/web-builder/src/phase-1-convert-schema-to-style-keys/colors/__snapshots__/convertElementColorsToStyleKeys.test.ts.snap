// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertElementColorsToStyleKeys > generates style keys for multiple palette entries 1`] = `
{
  "p1": {
    "danger": {
      "focus": [
        "borderColor==focus__[10,20,30,0.1]",
      ],
      "rest": [
        "borderColor--rest__[0,0,0,0.02]",
      ],
    },
    "primary": {
      "hover": [
        "textColor==hover__[240,50,50,0.5]",
      ],
      "rest": [
        "textColor--rest__[120,50,50,1]",
        "borderColor--rest__[120,50,50,1]",
      ],
    },
    "secondary": {
      "rest": [
        "textColor--rest__[240,50,50,0.5]",
      ],
    },
  },
}
`;

exports[`convertElementColorsToStyleKeys > generates style keys for palette property with a reference value 1`] = `
{
  "p1": {
    "primary": {
      "hover": [
        "borderColor==hover__[255,255,255,0.1]",
      ],
      "rest": [
        "borderColor--rest__[255,255,255,1]",
      ],
    },
  },
}
`;

exports[`convertElementColorsToStyleKeys > generates style keys for palette property without reference 1`] = `
{
  "p1": {
    "primary": {
      "rest": [
        "boxColor--rest__[45,100,50,1]",
      ],
    },
  },
}
`;

exports[`convertElementColorsToStyleKeys > includes parent style keys when only reference interaction states are provided 1`] = `
{
  "element": {
    "p1": {
      "primary": {
        "focus": [
          "boxColor==focus__[40,50,60,0.5]",
        ],
        "hover": [
          "boxColor==hover__[10,20,30,1]",
        ],
        "rest": [
          "boxColor--rest__[0,128,255,1]",
        ],
      },
    },
  },
  "parent": [
    "==hover",
    "==focus",
  ],
}
`;

exports[`convertElementColorsToStyleKeys > treats direct interaction-state map as neutral semantic color 1`] = `
{
  "p1": {
    "neutral": {
      "hover": [
        "boxColor==hover__[0,128,255,0.5]",
      ],
      "rest": [
        "boxColor--rest__[0,128,255,1]",
      ],
    },
  },
}
`;
